charms/istio-ingressgateway/src/charm.py

```diff
@@ -0,0 +1,229 @@
+#!/usr/bin/env python3
+
+import logging
+from pathlib import Path
+import os
+from kubernetes import client, config
+
+from ops.charm import CharmBase
+from ops.main import main
+from ops.model import ActiveStatus, MaintenanceStatus, WaitingStatus, BlockedStatus
+
+from oci_image import OCIImageResource, OCIImageResourceError
+from k8s_service import RequireK8sService
+
+
+class IstioIngressGatewayCharm(CharmBase):
+    def __init__(self, *args):
+        super().__init__(*args)
+        if not self.unit.is_leader():
+            # We can't do anything useful when not the leader, so do nothing.
+            self.model.unit.status = WaitingStatus("Waiting for leadership")
+            return
+        self.log = logging.getLogger(__name__)
+        self.istio_pilot = RequireK8sService(self, "istio-pilot")
+        self.image = OCIImageResource(self, "oci-image")
+        for event in [
+            self.on.install,
+            self.on.leader_elected,
+            self.on.upgrade_charm,
+            self.on.update_status,
+            self.on.config_changed,
+            self.istio_pilot.on.k8s_services_changed,
```

Comment:
Is this a typo, or is `k8s-service` from metadata.yaml supposed to be pluralized here?

Reply:
Also, is there any reason not to just use `self.on['istio-pilot'].relation_changed` here? I'm not sure what's going with `RequireK8sService` TBH. Is that an operator framework thing, or something we made?

Reply:
This is an [interface](https://github.com/charmed-kubernetes/interface-k8s-service) we wrote some months ago for when charms need to connect to a service of another charm.

`k8s_services_changes` is a custom event of the interface which actually checks if the data has changed before sending out an event. It is pluralized because you can provide multiple services. You can checkout the code [here](https://github.com/charmed-kubernetes/interface-k8s-service/blob/main/k8s_service.py#L17).